sdk/core/azure-core/inc/azure/core/http/body_stream.hpp

```diff
@@ -37,6 +37,20 @@ namespace Azure { namespace Core { namespace Http {
    *@brief Used to read data to/from a service.
    */
   class BodyStream {
+  protected:
+    /**
+     * @brief Read portion of data into a buffer.
+     *
+     * @remark This is the Read implementation that all derived classes need to provide.
+     *
+     * @param conntext #Context so that operation can be canceled.
+     * @param buffer Pointer to a first byte of the byte buffer to read the data into.
+     * @param count Size of the buffer to read the data into.
+     *
+     * @return Number of bytes read.
+     */
+    virtual int64_t ReadImplementation(Context const& context, uint8_t* buffer, int64_t count) = 0;
```

Comment:
We should settle on a naming convention. `Implementation` vs `Internal` vs something else.

Also, if no one other than `BodyStream` is supposed to call this, make it `private`.

Reply:
This as a public API surface name, isn't great. We'd want most (if not all) folks to call `Read`, which is a lot more user friendly, and not call `ReadImplementation` directly.

Reply:
All folks will call `Read()`.  All folks are currently and will continue to call `Read()` only.

`ReadImplementation()` is a contract for anyone creating a new bodyStream to implement.   
In the past, the contract was to implement `Read()`, but now `Read()` should not be overridden.

Reply:
I think we should rename ReadImplementation to OnRead as this is a common OOP pattern.
@CaseyCarter: is this OnXxx pattern common in C++?

Reply:
updated to OnRead(). I liked it

Reply:
We should add this in the guidelines so we are consistent with `Operation<T>` and other usages of private/protected virtual methods. Either we do: `OnX` or `XImplementation`, so `OnPollUntilDone`?